Add a maximal response size parameter for the HTTP Request feature (#48)
dfinity/interface-spec: spec/index.adoc

For security reasons, only HTTPS connections are allowed (URLs must start with `https://`). The IC uses industry-standard root CA lists to validate certificates of remote web servers.

> Each request can specify the maximal expected size for the response. The upper limit on the size of a response is `2MiB`.
> If the response is larger than the specified size, or than `2MiB` if no maxmial size is specified, an error will be returned.

Reviewer: I would simplify the third sentence. 
 
 ```suggestion
Each request can specify the maximal expected size for the response. The upper limit on the size of a response is `2MiB` if no maximal size value is specified.
An error will be returned when the response is larger than the maximal size. 
```

The maximal size of a request URL is 2048 bytes.

> Each request should specify the maximal expected size for the response. The maximal size of a response is `2MiB`. Therefore, only the first number of bytes as specified in the request (up to 2MiB), will be returned if a response is larger than these limits. The 2MiB size limit also applies to the value returned by the `transform` function.

Reviewer 2: Silent truncation sounds like a footgun.

Can’t we return an error in that case? HTTP error 413 isn’t quite the right thing, because that seems to be about the request size. Maybe 206 Partial Content?

Or maybe better to issue an IC-level reject here.

Author: This is not that silent if the user specifies the maximal size.
The idea is to encourage users to send requests that result in smaller responses, as we are worried that this feature could use lots of bandwidth otherwise. I do see cases where the first X bytes of a bigger response are important.
We do not want to reject the call if the first X requested bytes can be served..

Reviewer 2: But it is silent! How do I know whether the request was truncated, or just happened to have the size I was expecting? And if the user does not specify the size explicitly, it’s still 2MiB, right?

>  We do not want to reject the call if the first X requested bytes can be served..

Would that use-case be better serves using the `Ranges:` header of HTTP? This is an existing protocol lettings users request parts of a response?

We could either hope that most servers support this, or implement that ranges in the HTTP component.

Or at least add an `IC-response-truncated:` header maybe? Still a footgun, I’d say.

Reviewer 3: As an engineer, I see little usage in receiving only part of a response, tbh. Thus, I agree that we should fail hard—and early.

Author: Users could use ranges in their request.
I meant that it is not silent as the user specifies the size and the response should have a `Content-Length` header so they can know if it was truncated. But I also see the un-usefulness in this kind of truncated response...

Reviewer 2: Is that header guaranteed to be there? If the server streams responses probably not.

Reviewer 3: Indeed, if `Transfer-Encoding` is set, `Content-Length` must not be set ...

https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html

While for most common use cases, Content-Length is set—there are sloppy http implementations out there.

Reviewer 4: I agree with @nomeata's concern. We should not discard silently.

Reviewer: I also think it's better to reject if the response does not fit the limit provided. Having this parameter that the users can set makes them more aware of the limitations.

Author: Please see updated text

> Each request should specify the maximal expected size for the response. The upper limit on the size of a response is `2MiB`. Therefore, only the first number of bytes as specified in the request (up to 2MiB), will be returned if a response is larger than these limits. The 2MiB size limit also applies to the value returned by the `transform` function.

Reviewer 5: Shall we specify the unit expected for this field? It feels like `KB` is the intended unit, in that case, we probably want to explain the upper limit in targeted unit.

Reviewer 3: ... the way I read it, the unit is `bytes`.

Reviewer 3: But, yes, we should make it more explicit.

Author: please see updated text
